In [1]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeableNote: you may need to restart the kernel to use updated packages.




[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC13\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [2]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC13\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [3]:
pip install matplotlib

Defaulting to user installation because normal site-packages is not writeableNote: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC13\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [4]:
pip install seaborn

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC13\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [5]:
pip install scipy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB5-StudentPC13\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [6]:
import numpy as np

In [7]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats 
from scipy.stats import f_oneway

<h1>DATA ANALYTICS PROJECT</h1>

<H2>PROJECT: UNITED KINGDOM ROAD ACCIDENT DATA ANALYSIS</H2>
<H2>INCLUSIVE YEARS: 2019 - 2022</H2>
<H3>ANALYST: MARC CHRISTIAN MACARAIG</H3>

In [8]:
accident = pd.read_csv('dataset\\accident_data.csv')

In [9]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [10]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1>DATA CLEANSING</h1>

In [11]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [12]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unaccounted')
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna('unaccounted')

In [13]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [14]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [15]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [16]:
accident.dtypes

Index                        object
Accident_Severity          category
Accident Date                object
Latitude                    float64
Light_Conditions             object
District Area                object
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions      object
Road_Type                    object
Urban_or_Rural_Area          object
Weather_Conditions           object
Vehicle_Type                 object
dtype: object

<h1>CONVERTING OBJECT TO DATETIME DATA TYPE</h1>

In [17]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [18]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [19]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

In [20]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  object        
 5   District Area            660679 non-null  object        
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  object        
 10  Road_Type                660679 non-null  object        
 11  Urban_or_Rural_Area      660679 non-null  object        
 12  Weather_Conditio

<h1>INSIGHT 1</h1>
<H1>What proportion of accidents fall under each severity level?</H1>

In [21]:
severity_distribution = accident['Accident_Severity'].value_counts(normalize=True) * 100
print("Accident Severity Distribution:\n", severity_distribution)

Accident Severity Distribution:
 Accident_Severity
Slight     85.336601
Serious    13.352475
Fatal       1.310924
Name: proportion, dtype: float64


<h3>-The result shows that among the three accident severity categories, "Slight Accidents" account for the highest proportion at 85.33%, indicating that minor collisions (e.g., fender benders, low-speed impacts) are the most frequent type of road incident.</h3>

<h1>INSIGHT 2</h1>
<H1>EXCTRACTING DATA INFORMATION USING PANDAS DATE TIME</H1>

In [22]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [23]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

<h1>INSIGHT 3</h1>
<H1>Which light conditions are most frequently associated with accidents?</H1>

In [24]:
light_conditions = accident['Light_Conditions'].value_counts()
print("Most Common Light Conditions:\n", light_conditions)

Most Common Light Conditions:
 Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64


<h3>-The result shows that daylight conditions account for the highest number of accidents, with a total of 484,880 incidents recorded.</h3>

<h1>INSIGHT 4</h1>
<H1> What percentage of accidents occur in urban vs. rural areas?<H1>

In [25]:
urban_rural = accident['Urban_or_Rural_Area'].value_counts(normalize=True) * 100
print("Urban vs Rural Accidents:\n", urban_rural)

Urban vs Rural Accidents:
 Urban_or_Rural_Area
Urban          63.822673
Rural          36.173391
unaccounted     0.002270
Unallocated     0.001665
Name: proportion, dtype: float64


<h3>-The result shows that urban areas account for 63.82% of accidents, compared to 36.17% in rural areas, indicating that urban regions are significantly more prone to accidents. </h3>

<h1>INSIGHT 5</h1>
<H1>Which weather conditions are most associated with accidents?</H1>

In [26]:
weather_conditions = accident['Weather_Conditions'].value_counts()
print("Accidents by Weather Conditions:\n", weather_conditions)

Accidents by Weather Conditions:
 Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64


<h3>-The result shows that among weather conditions, "Fine weather with no high winds" leads in accident frequency, while "snowing with high winds" has the lowest accident rate.</h3>

<h1>INSIGHT 6</h1>
<H1>Which road surface conditions are most common during accidents?</H1>

In [27]:
road_surface = accident['Road_Surface_Conditions'].value_counts()
print("Accidents by Road Surface Conditions:\n", road_surface)

Accidents by Road Surface Conditions:
 Road_Surface_Conditions
Dry                     447821
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
unaccounted                726
Name: count, dtype: int64


<h3>-The result shows that dry road surfaces are the most prone to accidents, which may initially seem counterintuitive given common assumptions that wet, icy, or slippery conditions pose greater risks. </h3>

<h1>INSIGHT 7</h1>
<H1>Which district areas have the highest number of accidents?</H1>

In [28]:
district_accidents = accident.groupby('District Area').size().sort_values(ascending=False)
print("Top 10 Most Dangerous District Areas:\n", district_accidents.head(10))

Top 10 Most Dangerous District Areas:
 District Area
Birmingham          13491
Leeds                8898
Manchester           6720
Bradford             6212
Sheffield            5710
Westminster          5706
Liverpool            5587
Glasgow City         4942
Bristol, City of     4819
Kirklees             4690
dtype: int64


<h3>-The result shows that the Birmingham area is the most dangerous in terms of accident frequency, with a total of 13,491 recorded accidents.</h3>

<h1>INSIGHT 8</h1>
<H1>Which vehicle types are most frequently involved in accidents?</H1>

In [29]:
vehicle_types = accident['Vehicle_Type'].value_counts()
print("Vehicle Types Involved in Accidents:\n", vehicle_types)

Vehicle Types Involved in Accidents:
 Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64


<h3>-The result shows that cars are the most common vehicle type involved in accidents, which aligns with their prevalence on roads and high usage rates.</h3>

<h1>INSIGHT 9</h1>
<H1>What is the average number of casualties per accident?</H1>

In [30]:
avg_casualties = accident['Number_of_Casualties'].mean()
print(f"Average Number of Casualties per Accident: {avg_casualties:.2f}")

Average Number of Casualties per Accident: 1.36


<H3>-The result shows that the average number of casualties per accident is 1.36, indicating that each collision typically results in slightly more than one injury or fatality.</H3>

<h1>INSIGHT 10</h1>
<H1> Is there a correlation between the number of vehicles and casualties?</H1>

In [31]:
correlation = accident[['Number_of_Vehicles', 'Number_of_Casualties']].corr().iloc[0, 1]
print(f"Correlation Between Vehicles and Casualties: {correlation:.2f}")

Correlation Between Vehicles and Casualties: 0.23


<h3>-The result shows that the correlation between the number of vehicles and casualties is 0.23, indicating a weak positive relationship</H3>

<h1>INSIGHT 11</h1>
<H1>Which road types have the highest accident rates?</H1>

In [32]:
road_type_counts = accident['Road_Type'].value_counts()
print("Accidents by Road Type:\n", road_type_counts)

Accidents by Road Type:
 Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unaccounted             4520
Name: count, dtype: int64


<H3>-The result shows that among all road types, single carriageways are more prone to accidents.</H3>

<h1>INSIGHT 12</h1>
<H1>What are the latitude and longitude ranges for accidents?</H1>

In [33]:
geo_summary = accident[['Latitude', 'Longitude']].describe()
print("Geographical Summary:\n", geo_summary)

Geographical Summary:
             Latitude      Longitude
count  660679.000000  660679.000000
mean       52.553862      -1.431193
std         1.406896       1.383305
min        49.914430      -7.516225
25%        51.490696      -2.332191
50%        52.315755      -1.411613
75%        53.453444      -0.232901
max        60.757544       1.762010


<h1>INSIGHT 13</h1>
<H1>Are accidents more frequent during the day or night?</H1>

In [34]:
day_night = accident['Light_Conditions'].apply(lambda x: 'Day' if 'day' in x.lower() else 'Night')
day_night_counts = day_night.value_counts()
print("Day vs Night Accidents:\n", day_night_counts)

Day vs Night Accidents:
 Light_Conditions
Day      484880
Night    175799
Name: count, dtype: int64


<H3>-The result shows that accidents are more prone to occur during daytime hours compared to nighttime.</H3>

<h1>INSIGHT 14</h1>
<H1>Do accidents peak during certain seasons?</H1>

In [35]:
accident['Month'] = accident['Accident Date'].dt.month
seasons = {1: 'Winter', 2: 'Winter', 3: 'Spring', 4: 'Spring', 5: 'Spring',
           6: 'Summer', 7: 'Summer', 8: 'Summer', 9: 'Autumn', 10: 'Autumn',
           11: 'Autumn', 12: 'Winter'}
accident['Season'] = accident['Month'].map(seasons)
season_counts = accident['Season'].value_counts()
print("Accidents by Season:\n", season_counts)

Accidents by Season:
 Season
Autumn    176459
Summer    167839
Spring    162182
Winter    154199
Name: count, dtype: int64


<H3>-The result shows that among all four seasons, autumn has the highest rate of accidents.</H3>

<h1>INSIGHT 15</h1>
<H1>Identify potential causes by combining features.</H1>

In [36]:
top_causes = accident.groupby(['Weather_Conditions', 'Road_Surface_Conditions']).size().reset_index(name='Counts')
top_causes = top_causes.sort_values(by='Counts', ascending=False).head(10)
print("Top Causes of Accidents:\n", top_causes)

Top Causes of Accidents:
        Weather_Conditions Road_Surface_Conditions  Counts
6      Fine no high winds                     Dry  427799
10     Fine no high winds             Wet or damp   81178
33  Raining no high winds             Wet or damp   78087
45            unaccounted                     Dry   11223
8      Fine no high winds            Frost or ice   10653
22                  Other             Wet or damp    9366
28   Raining + high winds             Wet or damp    9184
0       Fine + high winds                     Dry    5023
20                  Other            Frost or ice    4718
43  Snowing no high winds                    Snow    3659


<h1>INSIGHT 16</h1>
<H1>What factors are most common in fatal accidents<H1>

In [37]:
fatal_accidents = accident[accident['Accident_Severity'] == 'Fatal']
fatal_summary = fatal_accidents.describe(include='all')
print("Summary of Fatal Accidents:\n", fatal_summary)

Summary of Fatal Accidents:
            Index Accident_Severity                  Accident Date  \
count       8661              8661                           8661   
unique      5095                 1                            NaN   
top     2.01E+12             Fatal                            NaN   
freq        3567              8661                            NaN   
mean         NaN               NaN  2020-10-14 09:03:40.713543680   
min          NaN               NaN            2019-01-01 00:00:00   
25%          NaN               NaN            2019-10-25 00:00:00   
50%          NaN               NaN            2020-09-18 00:00:00   
75%          NaN               NaN            2021-09-09 00:00:00   
max          NaN               NaN            2022-12-31 00:00:00   
std          NaN               NaN                            NaN   

           Latitude Light_Conditions District Area    Longitude  \
count   8661.000000             8661          8661  8661.000000   
unique  

<h1>INSIGHT 17</h1>
<H1>Which vehicle types result in the highest casualty rates?</H1>

In [38]:
casualty_by_vehicle = accident.groupby('Vehicle_Type')['Number_of_Casualties'].mean().sort_values(ascending=False)
print("Casualty Rate by Vehicle Type:\n", casualty_by_vehicle)

Casualty Rate by Vehicle Type:
 Vehicle_Type
Data missing or out of range             1.500000
Pedal cycle                              1.370558
Taxi/Private hire car                    1.368663
Goods over 3.5t. and under 7.5t          1.362861
Motorcycle over 500cc                    1.359434
Car                                      1.358841
Van / Goods 3.5 tonnes mgw or under      1.354537
Goods 7.5 tonnes mgw and over            1.351881
Bus or coach (17 or more pass seats)     1.349216
Minibus (8 - 16 passenger seats)         1.345648
Motorcycle over 125cc and up to 500cc    1.343521
Agricultural vehicle                     1.342065
Other vehicle                            1.340075
Motorcycle 50cc and under                1.337235
Motorcycle 125cc and under               1.332635
Ridden horse                             1.250000
Name: Number_of_Casualties, dtype: float64


<H3>-The result shows that the vehicle type categorized as "Data missing or out of range" accounted for the highest casualty rate, with a percentage of 1.500000</H3>

<h1>INSIGHT 18</h1>
<H1>Are accidents more frequent on weekends or weekdays?</H1>

In [39]:
accident['Day_of_Week'] = accident['Accident Date'].dt.day_name()
weekend_weekday = accident['Day_of_Week'].apply(lambda x: 'Weekend' if x in ['Saturday', 'Sunday'] else 'Weekday')
weekend_weekday_counts = weekend_weekday.value_counts()
print("Weekend vs Weekday Accidents:\n", weekend_weekday_counts)

Weekend vs Weekday Accidents:
 Day_of_Week
Weekday    464199
Weekend    196480
Name: count, dtype: int64


<H3>-The result shows that weekdays are associated with higher accident rates compared to weekends.</H3>

<h1>INSIGHT 19</h1>
<H1> Identify geographical hotspots based on latitude and longitude.</H1>

In [40]:
accident['Lat_Bin'] = pd.cut(accident['Latitude'], bins=10)
accident['Lon_Bin'] = pd.cut(accident['Longitude'], bins=10)
hotspots = accident.groupby(['Lat_Bin', 'Lon_Bin']).size().reset_index(name='Counts')
hotspots = hotspots.sort_values(by='Counts', ascending=False).head(10)
print("Top 10 Accident Hotspots:\n", hotspots)

Top 10 Accident Hotspots:
              Lat_Bin            Lon_Bin  Counts
17  (50.999, 52.083]  (-1.021, -0.0936]  106119
18  (50.999, 52.083]   (-0.0936, 0.834]   71839
26  (52.083, 53.167]   (-1.949, -1.021]   57684
35  (53.167, 54.252]   (-2.877, -1.949]   55517
36  (53.167, 54.252]   (-1.949, -1.021]   50215
25  (52.083, 53.167]   (-2.877, -1.949]   33614
46  (54.252, 55.336]   (-1.949, -1.021]   25901
15  (50.999, 52.083]   (-2.877, -1.949]   21414
34  (53.167, 54.252]   (-3.805, -2.877]   20456
27  (52.083, 53.167]  (-1.021, -0.0936]   18333


C:\Users\LAB5-StudentPC13\AppData\Local\Temp\ipykernel_2732\3595062550.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  hotspots = accident.groupby(['Lat_Bin', 'Lon_Bin']).size().reset_index(name='Counts')


<h1>INSIGHT 20</h1>
<H1>Most dangerous weather condition for accidents</H1>

In [41]:
accident.groupby("Weather_Conditions")["Accident_Severity"].value_counts()

Weather_Conditions     Accident_Severity
Fine + high winds      Slight                 7134
                       Serious                1245
                       Fatal                   175
Fine no high winds     Slight               441739
                       Serious               72046
                       Fatal                  7100
Fog or mist            Slight                 2963
                       Serious                 483
                       Fatal                    82
Other                  Slight                15184
                       Serious                1801
                       Fatal                   165
Raining + high winds   Slight                 8209
                       Serious                1261
                       Fatal                   145
Raining no high winds  Slight                69380
                       Serious                9468
                       Fatal                   848
Snowing + high winds   Slight            

<H3>The result shows that among all weather conditions, incidents of "fine weather with no high winds" were associated with the highest number of accidents.</H3>

<h1>DATA ANALYTICS</h1>
<H1>ANALYZING EACH FIELD FROM THE DATA SET</H1>

<H1>UniVariate</H1>

In [42]:
df_2019 = accident[accident['Year'] == 2019]

<h1>UniVariate Analysis from year 2019</h1>
<h1>Question: How many incidents happend under light conditions?</h1>

In [43]:
df_2019['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

In [44]:
df_2019['Light_Conditions'].value_counts().sum()

np.int64(182115)

<h1>Daylight Constitute to 132% of the total accidents from year 2019</h1>

<H1>BiVariate</H1>

<H1>MultiVariate</H1>